### PR TITLE
Issue-470-throttling

### PR DESCRIPTION
**Pull request by Goldziher**
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.md`?
- [x] Have you got 100% test coverage on new code?
- [x] Have you updated the prose documentation?
- [x] Have you updated the reference documentation?

This PR is a draft adding throttling #470

### CODE REVIEW

File: starlite/middleware/rate_limit.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    ) -> None:
+        """
+
+        Args:
+            app: The 'next' ASGI app to call.
+            rate_limit: A tuple containing a time unit.
+            exclude: A pattern or list of patterns to skip in the authentication middleware.
+        """
+        self.app = app
+        self.rate_limit_unit: DurationUnit = rate_limit[0]
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/middleware/rate_limit.py:83]: You can put different middlewares on different routes and/or combine this with exclusions. 

For example, you could set a throttle on a particular router, or just some route handlers.

**Review by cofin — APPROVED** [review]

**Comment by peterschutt** [comment]
I don't have a lot of experience with building out this sort of thing, but knowing I'm deploying applications with a redis cache available, the algorithm used here wouldn't be what I'd go for if I was to implement myself. E.g., maybe I'd try to use something like this: https://redis.com/redis-best-practices/basic-rate-limiting/

Do you intend for this to be a customizable implementation, or rather a quick "reach-for" type solution and rely on external library implementations for more targeted use?

Another thought I had was I might want to override the cache key function, perhaps to include a user key so the limiting wasn't host based, but user account based - would be easy to do by sub-classing the middleware, but the config hard-codes use of the middleware in the `middleware` property. Perhaps the middleware class could also be configurable on config object?

**Review by infohash — APPROVED** [review]

**Comment by pemocarlo** [comment]
I have also used this library with fastapi, I think it does something similar
https://github.com/laurentS/slowapi

**Comment by Goldziher** [comment]
> I don't have a lot of experience with building out this sort of thing, but knowing I'm deploying applications with a redis cache available, the algorithm used here wouldn't be what I'd go for if I was to implement myself. E.g., maybe I'd try to use something like this: https://redis.com/redis-best-practices/basic-rate-limiting/

The implementation needs to be agnostic to the cache backend used. Is there a particular issue with the algorithm used in your view?

Notes:

1. we need to be able to determine not only whether a response is allowed, but also how many requests are allowed within the given time frame (`RateLimit-Remaining`) and when the time frame will reset (`RateLimit-Reset`). Simply using cache expiration is not a solution for this use case. See the ietf specs: https://www.ietf.org/id/draft-ietf-httpapi-ratelimit-headers-05.html

> Do you intend for this to be a customizable implementation, or rather a quick "reach-for" type solution and rely on external library implementations for more targeted use?

The middleware is meant to be extensibe through inheritence. Additionally the user can pass in a throttle checker function to customize checks + configure the setting of headers.

> Another thought I had was I might want to override the cache key function, perhaps to include a user key so the limiting wasn't host based, but user account based - would be easy to do by sub-classing the middleware, but the config hard-codes use of the middleware in the `middleware` property. Perhaps the middleware class could also be configurable on config object?

Yup, I agree.

**Comment by peterschutt** [comment]
> The implementation needs to be agnostic to the cache backend used. Is there a particular issue with the algorithm used in your view?

No explicit issue, and given it has to work with any backend I have no problem. Like I mentioned, I'm no expert on this stuff, but just couldn't see how one could easily subclass or replace the middleware implementation if desired and that last point about making the middleware type configurable on the config object would sort that out.

**Comment by peterschutt** [comment]
I had a quick glance over a few throttling libraries last night, and there seem to be plenty of different strategies used across them. Overtime we could perhaps offer different strategies as alternate middlewares that could be set on the config object.

**Review by peterschutt — APPROVED** [review]

**Pull request merged**